### ISSUE TITLE
写给.NET程序员们：你所不知道的Python社区

### ISSUE THREAD

**Issue by tonyqus** [shnug/jobs#72]
可能关注我的粉丝最近会发现我时不时会转发一些Python相关的帖子，没错，我其实已经混Python社区有2年了，虽然是part-time混，不是全职混，毕竟我还是.NET社区的一员。

可能近几年，随着AI的普及，你们会发现Python这语言特别吃香，但是到底吃香到什么程度，已经Python语言在AI领域到底有多牛逼，你们可能还真不知道。

### 一些微软自己出的著名开源项目

[Autogen](https://github.com/microsoft/autogen) <img alt="Stars" src="https://img.shields.io/github/stars/microsoft/autogen?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Autogen)](https://pepy.tech/project/Autogen) -  Python实现，Agent自动化工具
[MarkItDown](https://github.com/microsoft/markitdown) <img alt="Stars" src="https://img.shields.io/github/stars/microsoft/markitdown?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/MarkItDown)](https://pepy.tech/project/MarkItDown) - Python实现，能将所有文件格式转换成markdown的库
[Mars](https://github.com/microsoft/MarS) <img alt="Stars" src="https://img.shields.io/github/stars/microsoft/MarS?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/MarS)](https://pepy.tech/project/MarS)  -  Python实现，微软亚洲研究院出品的量化交易模型
[Dapr](https://github.com/dapr/dapr) <img alt="Stars" src="https://img.shields.io/github/stars/dapr/dapr?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/dapr)](https://pepy.tech/project/dapr) - Golang实现，云原生架构框架
[Graphrag](https://github.com/microsoft/graphrag) <img alt="Stars" src="https://img.shields.io/github/stars/microsoft/graphrag?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Graphrag)](https://pepy.tech/project/Graphrag) - Python实现，图的检索增强方法

你想过，为什么这些项目微软程序员是用Python和Golang写， 而不是用.NET吗？不是说.NET才是微软亲儿子吗？其实任何大厂从来都不是傻子，他们很清楚没有任何一种编程语言能够通吃市场，即使这个语言是他们自己公司创造的。

在微软，不同团队的职责不一样，非.NET团队没有必要一味的去舔.NET组，他们会中立地去选择需要使用的编程语言。这就是他们选择其他语言的根本原因。

当然这些项目并不意味着微软会放弃.NET，只是微软并没有把鸡蛋都放在.NET一个篮子里而已。

### 一些非常有用且著名的Data Science库

以下这个库都是基于Python实现的用于Data Science的库，且下载量都是以G计算，说明其使用范围之广。Data Science不仅仅包含数据分析、可视化，还包含机器学习。

- NumPy <img alt="Stars" src="https://img.shields.io/github/stars/numpy/numpy?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/NumPy)](https://pepy.tech/project/NumPy) 支持一系列高等数学、线性代数函数，并支持多维数组
- Pandas <img alt="Stars" src="https://img.shields.io/github/stars/pandas-dev/pandas?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Pandas)](https://pepy.tech/project/Pandas) 可用于处理和操作表格数据，从而高效对数据集执行加载、对齐、合并和转换等操作
- Matplotlib <img alt="Stars" src="https://img.shields.io/github/stars/matplotlib/matplotlib?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Matplotlib)](https://pepy.tech/project/Matplotlib) 2D绘图库，能够以多种硬拷贝格式和跨平台的交互式环境生成出版物质量的图形
- Scikit-learn <img alt="Stars" src="https://img.shields.io/github/stars/scikit-learn/scikit-learn?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Scikit-learn)](https://pepy.tech/project/Scikit-learn) 各行各业广泛应用于机器学习和数据分析的通用工具。
- Seaborn <img alt="Stars" src="https://img.shields.io/github/stars/mwaskom/seaborn?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Seaborn)](https://pepy.tech/project/Seaborn)建立在Matplotlib 基础之上的Python 数据可视化库，专注于绘制各种统计图形
- Tensorflow <img alt="Stars" src="https://img.shields.io/github/stars/tensorflow/tensorflow?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/Tensorflow)](https://pepy.tech/project/Tensorflow) 使用数据流图形进行机器学习
- SciPy <img alt="Stars" src="https://img.shields.io/github/stars/scipy/scipy?style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/SciPy)](https://pepy.tech/project/SciPy) 算法库和数学工具包

### 语言性能真的这么重要吗？
有很多.NET程序员骂Python性能差，当然我也承认Python的性能确实不好，特别是Web领域，一直被吐槽无法用于高并发场景，即使有了FastAPI这样号称很快的Web框架，依然很慢。

但这真会影响语言在市场上的地位吗？为什么Python常年语言排名第一位呢？因为每一种语言有他专注的地方，Python就非常专注于自己擅长的领域——Data Science，而大多数情况下，数据分析师对于分析语言的性能并没有太多要求，主要就是他自己用，也不需要性能太好，能出图表和曲线就行。

.NET在最近几年一直尝试进入机器学习和人工智能领域，比较有名的是ML.NET，但这个库其实nuget下载量并不高。再来看最近MVP们一直在推的Semantic Kernel与其竞品Langchain的对比。

- Semantic Kernel <img alt="Github Stars" src="https://img.shields.io/github/stars/microsoft/semantic-kernel?logo=github&style=flat-square&labelColor=343b41"/> [![Pip Downloads](https://static.pepy.tech/badge/semantic-kernel)](https://pepy.tech/project/semantic-kernel) [![NuGet](https://img.shields.io/nuget/dt/Microsoft.SemanticKernel?label=Nuget-downloads)](https://www.nuget.org/packages/Microsoft.SemanticKernel)
- LangChain <img alt="Github Stars" src="https://img.shields.io/github/stars/langchain-ai/langchain?logo=github&style=flat-square&labelColor=343b41"/> [![pip-downloads](https://static.pepy.tech/badge/langchain)](https://pepy.tech/project/langchain) ![npm](https://img.shields.io/npm/dt/langchain?label=npm-downloads) 

其实人还是要知难而退的，当你明知道这条路走不通，你就要另辟蹊径，就这么简单。其实.NET可以专注于自己的领域，比如工控、医疗、金融、航空，没必要去抢数据科学领域，那个水是很深的。同样的道理，.NET也没必要用Blazor去抢前端的生意，最后倒霉的还是那些无知的程序员，跟着微软学了一些没有用的东西，最后微软把这些项目叫停了。

### Reference
https://www.datacamp.com/blog/top-python-libraries-for-data-science
https://github.com/tonyqus/langchain-vs-semantic-kernel

**Comment by wuuer** [comment]
.NET 金融领域使用多吗?，我看到的都是C++或Python居多

**Comment by tonyqus** [comment]
有几个大的投行在用，比如JP Morgan，高盛，花旗。但都是前端项目，数据分析依然是Python，量化考虑到性能，大概率是C++和Rust。
